File: 935-knight-dialer/935-knight-dialer.cpp
class Solution {
public:
    
    
    
    int knightDialer(int n) {
        int dp[n+1][10];
        long long int mod = 1000000007;
        memset(dp, 0, sizeof(dp));
        int i, j, k;
        for(i=0;i<10;i++){
            dp[1][i] = 1;
        }
        
        map<int, vector<int>> m;
        m[0] = {4,6};
        m[1] = {6,8};
        m[2] = {7,9};
        m[3] = {4,8};
        m[4] = {3,9, 0};
        m[5] = {};
        m[6] = {7,1, 0};
        m[7] = {2,6};
        m[8] = {1,3};
        m[9]= {2,4};
        
    
        for(i=0;i<10;i++){
            dp[1][i] = 1;
        }
        
        for (i=2;i<=n;i++){
            for(j=0;j<10;j++){
                for(k =0;k<m[j].size();k++){
                    // cout<<j<<" "<<m[j][k]<<"\n";
                    dp[i][j] = (dp[i][j]%mod + dp[i-1][m[j][k]]%mod) %mod;
                }
            }
        }
        long long int total = 0;
        for(i=0;i<10;i++){
            // cout<<i<<" "<<dp[n][i]<<"\n";
            total += dp[n][i]%mod ;
        }
        
        return (total%mod);
        
        
    }
};

File: 935-knight-dialer/935-knight-dialer.py
class Solution:
    
    def recurse(self, m, n):
        mod = 1000000007

        if(n==1):
            return [1,1,1,1,1,1,1,1,1,1]
        val = [0]*10
        arr = self.recurse(m, n-1)
        for i in range(0, 10):
            for k in m[i]:
                val[k] = (val[k]%mod + arr[i]%mod)%mod
        return val
    
    def knightDialer(self, n: int) -> int:
        # dp = [[0]*(10)]*(n+1)
        mod = 1000000007
        # print("After initialization", dp)
        
        stepsMap= {
            1: [6,8],
            2: [7,9],
            3: [4,8],
            4: [3,9,0],
            5: [],
            6: [7,1,0],
            7: [2,6],
            8: [1,3],
            9: [4,2],
            0: [4,6],
        }
        
        dp = self.recurse(stepsMap, n)
        # print(dp)
        total = 0
        for i in range(0, 10):
            # print(dp[i])
            total += dp[i]%mod
        return total%mod
            